### PR TITLE
Try harder to Delete Files on Windows

### LINKED ISSUE #25984: Provide better diagnostics for the "Unable to delete directory... Failed to delete some children..." error
### Expected Behavior

The "Unable to delete directory... Failed to delete some children..." error thrown by Gradle should be more helpful.

### Current Behavior

When running tests for the Android Gradle plugin, we occasionally run into the following flaky error:
```
Caused by: org.gradle.internal.exceptions.LocationAwareException: Execution failed for task ':app:packageRelease'.
  ...
Caused by: java.io.IOException: Unable to delete directory 'C:\...\project\app\build\intermediates\incremental\packageRelease\tmp'
  Failed to delete some children. This might happen because a process has files open or has its working directory set in the target directory.
  - C:\...\project\app\build\intermediates\incremental\packageRelease\tmp\release
	at org.gradle.internal.file.impl.DefaultDeleter.throwWithHelpMessage(DefaultDeleter.java:181)
	at org.gradle.internal.file.impl.DefaultDeleter.deleteRecursively(DefaultDeleter.java:114)
	at org.gradle.internal.file.impl.DefaultDeleter.ensureEmptyDirectory(DefaultDeleter.java:83)
	at org.gradle.internal.file.impl.DefaultDeleter.ensureEmptyDirectory(DefaultDeleter.java:75)
	at org.gradle.internal.execution.steps.RemovePreviousOutputsStep$2.visitOutputProperty(RemovePreviousOutputsStep.java:121)
```

Note that the task names and directory names are not relevant as they may change. Our focus is on the error message.

The last lines of the error message just before the stack trace show the regular files or directories that Gradle is unable to delete (`packageRelease\tmp\release` in the above example).
   - If it is a regular file, the user can investigate that file with tools such as https://file-leak-detector.kohsuke.org, or report the issue to tools that are not Gradle.
   - However, if it is a directory, the user is stuck. **The error message doesn't say why the directory can't be deleted**. 
       + The most likely root cause in this case is the infamous `DirectoryNotEmptyException` on Windows where a deleting an empty directory might fail the first time but succeed if we retry. Here are some examples in [Gradle](https://github.com/gradle/gradle/blob/3dc80197fdc09fbf36e909951bd95aac4b769bc3/subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java#L148), [Apache Flink](https://github.com/zentol/flink/blob/edbccbb5c28b83b469385d8683b6751810598246/flink-core/src/main/java/org/apache/flink/util/FileUtils.java#L179), and the [Android Gradle plugin](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-master-dev:common/src/main/java/com/android/utils/PathUtils.java?q=%22DirectoryNotEmptyException%20exception%22) where this retry logic is implemented.
       + It is possible to confirm this root cause by providing better diagnostics -- see below.
       
****Proposed Changes****       
I'm proposing 2 changes in Gradle to help with this error:
   1. In [DefaultDeleter](https://github.com/gradle/gradle/blob/3dc80197fdc09fbf36e909951bd95aac4b769bc3/subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java#L148), consider using "assertive" API such as `java.nio.file.Files.deleteIfExists` which throws instead of returning `false` when it can't delete a file/directory.
   2. In the retry logic of [DefaultDeleter](https://github.com/gradle/gradle/blob/3dc80197fdc09fbf36e909951bd95aac4b769bc3/subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java#L148), consider:
        - Adding more wait time between two retries (currently the wait time is 10ms)
        - Retrying a few more times (currently Gradle retries only once)
        - *(The [Android Gradle plugin](https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-master-dev:common/src/main/java/com/android/utils/PathUtils.java?q=%22DirectoryNotEmptyException%20exception%22) uses both strategies.)*       
        

### Context (optional)

_No response_

### Steps to Reproduce

N/A (flaky error)

### Gradle version

8.2 (and earlier)

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

### PR DESCRIPTION
**Pull request by nodmp**
* Fixes #25984 

### Context

on the Windows filesystem, such that if we delete a directory immediately after deleting its contents, the deletion might occasionally fail.

If a DirectoryNotEmptyException occurs when deleting, it may help to wait and try deleting a few more times

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] ~Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective~
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] ~Update User Guide, DSL Reference, and Javadoc for public-facing changes~
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by nodmp** [comment]
@eskatos, Can you review it for me? There's no one assigned here.

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java [hunk 1]
```diff
@@ -152,29 +158,40 @@ private boolean shouldRemoveContentsOf(File file, Handling handling) {
...
+                    Thread.sleep(DELETE_RETRY_SLEEP_MILLIS);
+                } catch (InterruptedException ex) {
+                    Thread.currentThread().interrupt();
+                }
+                try {
+                    result = deleteFile(file);
+                } catch (DirectoryNotEmptyException ex) {
+                    failedAttempts++;
+                    continue;
+                } catch (Exception ex) {
```
Comment (eskatos) [inline subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java:183]: ❓ I would only catch `IOException` here to let unexpected errors pop up

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java [hunk 2]
```diff
@@ -152,29 +158,40 @@ private boolean shouldRemoveContentsOf(File file, Handling handling) {
...
+                    return false;
+                }
+                break;
+            }
+            if (failedAttempts >= EMPTY_DIRECTORY_DELETION_ATTEMPTS) {
+                return false;
+            }
+        } catch (Exception exception) {
```
Comment (eskatos) [inline subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java:191]: ❓ I would only catch `IOException` here to let unexpected errors pop up

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java [hunk 3]
```diff
@@ -152,29 +158,40 @@ private boolean shouldRemoveContentsOf(File file, Handling handling) {
...
-        if (deleteFile(file)) {
-            return true;
-        }
-
-        // This is copied from Ant (see org.apache.tools.ant.util.FileUtils.tryHardToDelete).
-        // It mentions that there is a bug in the Windows JDK implementations that this is a valid
-        // workaround for. I've been unable to find a definitive reference to this bug.
-        // The thinking is that if this is good enough for Ant, it's good enough for us.
-        if (runGcOnFailedDelete) {
-            System.gc();
```
Comment (eskatos) [inline subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java:169]: ❌ This is an unneeded change of behaviour

I would keep `runGcOnFailedDelete` as is and not limit retries to Windows.

**Review by eskatos — CHANGES_REQUESTED** [review]
Thank you for the contribution.

Could you please add test coverage for the retry mechanism in `DefaultDeleterTest`?

I also requested a few changes, see below.

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java (hunk 1 above)
Comment (nodmp) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java]: Done

File: subprojects/files/src/test/groovy/org/gradle/internal/file/impl/DefaultDeleterTest.groovy [hunk 4]
```diff
@@ -188,6 +189,43 @@ class DefaultDeleterTest extends Specification {
...
+
+        and:
+        def ex = thrown IOException
+        normaliseLineSeparators(ex.message) == """
+            Unable to delete directory '$targetDir'
+              ${DefaultDeleter.HELP_FAILED_DELETE_CHILDREN}
+              - $nonDeletable
+        """.stripIndent().trim()
+    }
+
```
Comment (eskatos) [inline subprojects/files/src/test/groovy/org/gradle/internal/file/impl/DefaultDeleterTest.groovy:228]: 👍

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java (hunk 3 above)
Comment (eskatos) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java:169]: ❌ Please restore the handling of `runGcOnFailedDelete`

**Review by eskatos — CHANGES_REQUESTED** [review]
Test coverage 👍

Please restore the handling of `runGcOnFailedDelete`, see my comment.

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java (hunk 3 above)
Comment (nodmp) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java:169]: Thanks for the review.
Do I need to restore the comments here or what?

Comment (eskatos) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java:169]: Yes please

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@nodmp, reviewing again I amended your branch with a couple changes:
- restore retry in all failure cases, this was an unnecessary behavior change
- don't rely on exceptions, given the above this is not necessary and might be more performant

I also refined the GC comment with evidence that it is necessary.

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

**Review by eskatos — APPROVED** [review]

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Pull request merged**